R/parse.R

```diff
@@ -135,3 +138,47 @@ parse_check_consistent_dimensions_lhs <- function(eq, dat, call) {
     }
   }
 }
+
+find_ranks_in_expr <- function(expr, reduction = FALSE) {
+  ranks <- list()
+  if (is.symbol(expr)) {
+    return(ranks)
+  }
+  is_array <- any(expr[[1]] == "[")
+
+  if (is_array && !reduction) {
+    array_name <- expr[[2]]
+    array_rank <- length(expr) - 2
+    entry <- list()
+    entry[[array_name]] <- array_rank
+    ranks <- c(ranks, entry)
+  }
+
+  reduction <- expr[[1]] == "OdinReduce"
+
+  for (i in seq_along(expr)) {
+    e <- expr[[i]]
+    if (is.recursive(e)) {
+      ranks <- c(ranks, find_ranks_in_expr(e, reduction))
+    }
+  }
+  ranks
+}
+
+parse_check_consistent_dimensions_rhs <- function(eq, dat, call) {
+  arrays <- intersect(eq$rhs$depends$variables, dat$storage$arrays$name)
```

Comment:
I don't think this is what you want to do here - we're interested to see that non-arrays are **not** referenced as arrays too

Reply:
I thought that was already covered and threw an error in the expression parsing. So this one is dealing with inappropriate usage of arrays that we spot at the system level, but there is already a `dim` on some other line...?

Reply:
So here I'm speeding things up by not bothering to check if there's nothing on the RHS that we didn't previous note as having a `dim` for

Reply:
> I thought that was already covered and threw an error in the expression parsing

Does it? this stuff ends up hard to hold in one place!

Reply:
I will check again later! But I ... believe ... that here we are only dealing with the case of two things we know are both arrays, but rank might differ. The inner bracket of [], [,] [1,2,3] is not explored in the expression tests, but the outer type I think is. Errors of "it is/isn't an array" I think get thrown, but no rank comparison is done... (from memory...!)

Reply:
You're right (of course!) - this succeeds (pre this PR) - but won't get spotted with my current approach...!

```
odin_parse({
      initial(x) <- 0
      initial(a) <- 1
      update(x) <- a[1]
      update(a) <- 1
    })
```